### PR TITLE
Add DataFrameTransformer

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Addresses #501

### CODE REVIEW

File: skorch/helper.py [hunk 1]
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
...
+                "DataFrame contains a column named 'X', which clashes "
+                "with the name chosen for cardinal features; consider "
+                "renaming that column.")
+
+        wrong_dtypes = []
+
+        for col, dtype in zip(df, df.dtypes):
+            if isinstance(dtype, self.pd.core.dtypes.dtypes.CategoricalDtype):
+                continue
+            if dtype in self.int_dtypes_:
```
Comment (thomasjpfan) [inline skorch/helper.py]: Does `np.issubdtype(dtype, np.integer)` work?

File: skorch/helper.py [hunk 2]
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
...
+                continue
+            if dtype in self.float_dtypes_:
```
Comment (thomasjpfan) [inline skorch/helper.py]: Does `np.issubtype(dtype, np.floating)` work?

File: skorch/helper.py [hunk 3]
```diff
@@ -256,3 +261,268 @@ def predefined_split(dataset):
...
+      The dtype to cast the cardinal values to. If None, don't change
+      them.
+
+    int_dtype : numpy dtype or None (default=np.int64)
+      The dtype to cast the categorical values to. If None, don't
+      change them. If you do this, it can happen that the categorical
+      values will have different dtypes, reflecting the number of
+      unique categories.
+
+    signature_ : namedtuple
```
Comment (thomasjpfan) [inline skorch/helper.py]: A named tuple defined inside a class seems a little harder to discover. Would at tuple or a dictionary be enough?

**Comment by BenjaminBossan** [comment]
I changed the code to use `np.issubdtype` (learned something new). Moreover, I dropped the namedtuple but instead of using a tuple, I used a dict. The reason for that is the better repr, otherwise a lonely number is not very easy to interpret.

@thomasjpfan @ottonemo pls review again

**Comment by BenjaminBossan** [comment]
@thomasjpfan @ottonemo any things to add to this?

File: skorch/helper.py [hunk 4]
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
...
+    >>> # cast to category dtype to later learn embeddings
+    >>> df['col_cats'] = df['col_cats'].astype('category')
+    >>> y = np.asarray([0, 1, 0] * 4)
+
+    >>> class MyModule(nn.Module):
+    ...     def __init__(self):
+    ...         super().__init__()
+    ...         self.reset_params()
+
+    >>>     def reset_params(self):
```
Comment (thomasjpfan) [inline skorch/helper.py]: Should this be directly in `__init__`?

File: skorch/helper.py [hunk 5]
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
...
+        if 'X' in df:
+            raise ValueError(
+                "DataFrame contains a column named 'X', which clashes "
+                "with the name chosen for cardinal features; consider "
+                "renaming that column.")
+
+        wrong_dtypes = []
+
+        for col, dtype in zip(df, df.dtypes):
+            if isinstance(dtype, self.pd.core.dtypes.dtypes.CategoricalDtype):
```
Comment (thomasjpfan) [inline skorch/helper.py]: Does `pd.api.types.CategoricalDtype` work?

File: skorch/helper.py [hunk 6]
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
...
+          respective column names as keys.
+
+        """
+        self._check_dtypes(df)
+
+        X_dict = {}
+        Xf = []  # floats
+
+        for col, dtype in zip(df, df.dtypes):
+            values = df[col]
```
Comment (thomasjpfan) [inline skorch/helper.py]: Nit: Having `values.values` later is a bit harder to follow. Maybe `X_col=df[col`?

File: skorch/helper.py (hunk 4 above)
Comment (BenjaminBossan) [reply skorch/helper.py]: It's just my personal preference, that way it's possible to reset the module without creating a new instance.

File: skorch/helper.py (hunk 6 above)
Comment (BenjaminBossan) [reply skorch/helper.py]: Not a nit :+1:

**Comment by BenjaminBossan** [comment]
@ottonemo is this ready to be merged?

File: skorch/tests/test_helper.py [hunk 7]
```diff
@@ -474,3 +482,230 @@ def test_pickle(self, predefined_split, data):
...
+    @pytest.fixture
+    def df(self):
+        """DataFrame containing float, int, category types"""
+        import pandas as pd
+
+        df = pd.DataFrame({
+            'col_floats': [0.1, 0.2, 0.3],
+            'col_ints': [11, 11, 10],
+            'col_cats': ['a', 'b', 'a'],
+        })
```
Comment (thomasjpfan) [inline skorch/tests/test_helper.py]: I am a little surprised some of the test pass without this in python 3.5:

```suggestion
        }, columns=['col_floats', 'col_ints', 'col_cats'])
```

File: skorch/helper.py [hunk 8]
```diff
@@ -256,3 +261,250 @@ def predefined_split(dataset):
...
+
+            if isinstance(dtype, self.pd.api.types.CategoricalDtype):
+                x = X_col.cat.codes.values
+                if self.int_dtype is not None:
+                    x = x.astype(self.int_dtype)
+                X_dict[col] = x
+                continue
+
+            if (
+                    dtype in (np.dtype('int32'), np.dtype('int64'))
```
Comment (thomasjpfan) [inline skorch/helper.py]:
```suggestion
                    np.issubdtype(dtype, np.integer):
```

File: skorch/tests/test_helper.py (hunk 7 above)
Comment (BenjaminBossan) [reply skorch/tests/test_helper.py]: Why indicate the column names a second time?

**Comment by ottonemo** [comment]
I will merge this after the tests complete since this seems to be a version we can work with / build up on. What I would like to see in the future is a way to deal with NULL-values for categorical columns (introduce a separate category for NULL) since this is a common enough problem.

**Comment by thomasjpfan** [comment]
Since this PR uses the `cat.codes`, the `np.nan` values will be encoded as `-1`:

```py
import pandas as pd
df = pd.DataFrame({
    'col': pd.Categorical(['a', 'b', np.nan, 'b'])})

df['col'].cat.codes.values
# array([ 0,  1, -1,  1], dtype=int8)
```

**Pull request merged**